core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java

```diff
@@ -143,12 +170,15 @@ public void updateBufferSize(int bufferSize) throws TranslogException {
     /**
      * write all buffered ops to disk and fsync file
      */
-    public void sync() throws IOException {
+    public synchronized void sync() throws IOException { // synchronized to ensure only one sync happens a time
         // check if we really need to sync here...
         if (syncNeeded()) {
             try (ReleasableLock lock = writeLock.acquire()) {
-                lastSyncedOffset = writtenOffset;
-                checkpoint(lastSyncedOffset, operationCounter, channelReference);
+                ensureOpen();
+                final long offset = writtenOffset;
+                final int opsCount = operationCounter;
+                checkpoint(offset, opsCount, channelReference);
```

Comment:
Why do we need to assign to local `offset` and `opsCount` here?  Since we hold the write lock, `writtenOffset` cannot change between the call to `checkpoint` and when we assign it to `lastSyncedOffset`?